src/test/java/org/littleshoot/proxy/DirectRequestTest.java

```diff
@@ -0,0 +1,108 @@
+package org.littleshoot.proxy;
+
+import static org.hamcrest.Matchers.startsWith;
+import static org.junit.Assert.assertThat;
+import io.netty.handler.codec.http.DefaultHttpResponse;
+import io.netty.handler.codec.http.HttpObject;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpVersion;
+
+import java.net.Socket;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.littleshoot.proxy.impl.DefaultHttpProxyServer;
+import org.littleshoot.proxy.test.SocketClientUtil;
+
+/**
+ * This class tests direct requests to the proxy server, which causes endless
+ * loops (#205).
+ */
+public class DirectRequestTest {
+    private HttpProxyServer proxyServer;
+
+    private Socket socket;
+
+    @Before
+    public void setUp() throws Exception {
+        socket = null;
+        proxyServer = null;
+    }
+
+    @After
+    public void tearDown() throws Exception {
+        try {
+            if (proxyServer != null) {
+                proxyServer.abort();
+            }
+        } finally {
+            if (socket != null) {
+                socket.close();
+            }
+        }
+    }
+
+    @Test(timeout = 5000)
+    public void testDirectRequestAnsweredBadRequest() throws Exception {
+
+        HttpFiltersSource filtersSource = new HttpFiltersSourceAdapter() {
+            @Override
+            public HttpFilters filterRequest(HttpRequest originalRequest) {
+                return new HttpFiltersAdapter(originalRequest);
+            }
+        };
+
+        proxyServer = DefaultHttpProxyServer.bootstrap()//
+                .withPort(0)//
+                .withFiltersSource(filtersSource)//
+                .start();
+
+        socket = SocketClientUtil.getSocketToProxyServer(proxyServer);
```

Comment:
The reason I used the custom socket client for the KeepAliveTest is that I needed to verify the connection state. All the popular HTTP clients abstract connection management away from the client code, so I had to write a lower-level pseudo-client. The client still leaves a lot to be desired, however, and I would recommend using it only if other HTTP clients cannot properly test the functionality.

Regarding Apache HTTP Client, the way we are building it is deprecated, but the way we are using it is generally not deprecated. If we updated [AbstractProxyTest#buildHttpClient](https://github.com/adamfisk/LittleProxy/blob/master/src/test/java/org/littleshoot/proxy/AbstractProxyTest.java#L254) to use the 4.3+ builder, we could get rid of almost all of the deprecated HttpClient code.

I wouldn't say that URLConnection is the "regular" way to create an HTTPS client, but it is one way. It's certainly a fine HTTP client to use in unit tests. Since we're using Apache HttpClient in our other tests, my preference would be to continue to use it, unless there's some reason to do otherwise.